presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/Util.java

```diff
@@ -62,4 +64,23 @@ public static PlanNode transpose(PlanNode parent, PlanNode child)
                 parent.replaceChildren(
                         child.getSources())));
     }
+
+    /**
+     * @return an identity projection containing only those node outputs also in permittedOutputs, or the original node if all its inputs were permitted.
+     */
+    public static PlanNode restrictOutputs(PlanNodeIdAllocator idAllocator, PlanNode node, Set<Symbol> permittedOutputs)
+    {
+        List<Symbol> restrictedOutputs = node.getOutputSymbols().stream()
+                .filter(permittedOutputs::contains)
+                .collect(toImmutableList());
+
+        if (restrictedOutputs.size() < node.getOutputSymbols().size()) {
+            return new ProjectNode(
+                    idAllocator.getNextId(),
+                    node,
+                    Assignments.identity(restrictedOutputs)
```

Comment:
I think you should keep assignments as in former `PlanNode` .
E.g. in case of `node` being a `ProjectNode` with assignments of  `{ a_16 := a, b_12 := b + 5 }`. For `permittedOutputs` = `{a_16}`, your method would return Assignments of `{ a_16 := a_16 }`.
IMO you should have a special case for `ProjectNode` here (because `ProjectNode` has its own `Assignments`).
Previously it worked fine in context of `CrossJoinElimination` (if there were `ProjectNode` it'd be correct already but in case it's missing, we had to add it) but once you make it general-purpose util function, you should make sure it's correct in all cases.